logic/resolvable/Concludable.java

```diff
@@ -237,33 +239,39 @@ public FunctionalIterator<Unifier> unify(Rule.Conclusion.Relation relationConclu
                 }
             }
 
-            List<RolePlayer> conjRolePlayers = list(relation().players());
+            Set<RolePlayer> conjRolePlayers = relation().players();
             Set<RolePlayer> thenRolePlayers = relationConclusion.relation().players();
 
-            return matchRolePlayers(conjRolePlayers, thenRolePlayers, new HashMap<>(), conceptMgr)
+            return matchRolePlayers(conjRolePlayers, thenRolePlayers, conceptMgr)
                     .map(mapping -> convertRPMappingToUnifier(mapping, unifierBuilder.clone(), conceptMgr)).distinct();
         }
 
-        private FunctionalIterator<Map<RolePlayer, Set<RolePlayer>>> matchRolePlayers(
-                List<RolePlayer> conjRolePLayers, Set<RolePlayer> thenRolePlayers,
-                Map<RolePlayer, Set<RolePlayer>> mapping, ConceptManager conceptMgr) {
-            if (conjRolePLayers.isEmpty()) return single(mapping);
-            RolePlayer conjRP = conjRolePLayers.get(0);
-            return iterate(thenRolePlayers)
-                    .filter(thenRP -> iterate(mapping.values()).noneMatch(rolePlayers -> rolePlayers.contains(thenRP)))
-                    .filter(thenRP -> Unifier.Builder.unificationSatisfiable(conjRP, thenRP, conceptMgr))
-                    .map(thenRP -> {
-                        Map<RolePlayer, Set<RolePlayer>> clone = cloneMapping(mapping);
-                        clone.putIfAbsent(conjRP, new HashSet<>());
-                        clone.get(conjRP).add(thenRP);
-                        return clone;
-                    }).flatMap(newMapping -> matchRolePlayers(conjRolePLayers.subList(1, conjRolePLayers.size()),
-                                                              thenRolePlayers, newMapping, conceptMgr));
-        }
-
-        private Unifier convertRPMappingToUnifier(Map<RolePlayer, Set<RolePlayer>> mapping,
+        private FunctionalIterator<Map<RolePlayer, RolePlayer>> matchRolePlayers(Set<RolePlayer> conjRolePlayerSet, Set<RolePlayer> thenRolePlayers, ConceptManager conceptMgr) {
+            // TODO: If this is ever slow again, consider Divide & Conquer: Partition conjRolePlayers such that the result sets of all partitions are disjoint
+            // Sort, So that the once with identical role-labels are together; They hopefully fail together
+            List<RolePlayer> conjRolePlayers = new ArrayList<>(conjRolePlayerSet);
+            conjRolePlayers.sort(Comparator.comparing(rolePlayer -> rolePlayer.roleType().map(typeVariable -> typeVariable.label().map(LabelConstraint::label).orElse("")).orElse("")));
+            Map<RolePlayer, Set<RolePlayer>> unifiesWith = new HashMap<>();
+            iterate(conjRolePlayers).forEachRemaining(conjRP -> {
+                unifiesWith.put(conjRP, iterate(thenRolePlayers).filter(thenRP -> Unifier.Builder.unificationSatisfiable(conjRP, thenRP, conceptMgr)).toSet());
+            });
```

Comment:
Recursively (iteratively) construct the iterator, rather than doing matchRolePlayer recursion inside the iterator.

Much cleaner since the iterator all does the work, rather than interleaving iterator and the matchRolePlayer recursion.